### PR TITLE
Add support to dropout.

### PR DESCRIPTION
**Pull request by aliciafmachado**
* Add dropout.
* Add basic tests for dropout and trainer with dropout enabled.
* Add example in app.
* Add evalMode flag so that dropout can be disabled during eval.

Intended to resolve Issue: https://github.com/PAIR-code/tiny-transformers/issues/1

### CODE REVIEW

**Review by aliciafmachado — COMMENTED** [review]
There are some commits going back and forth on some things that I did not fully understood at first, so feel free to squash the commits before merging to main to avoid confusion. Otherwise I can recreate the pull request and fix the commit history.

I also have a few questions / discussion topics:
1. I added support to dropout but we need something to manage random seeds so that we can seed properly. Should we create an issue for that?
2. I tried to add dropout based on T5 architecture, but I decided to not add it after the FF layer and in the output. For the FF, I don't think it makes sense since we have a single layer and we apply dropout before the residual connection after the FF network. For the output, I don't see any additional computations after getting out of the stack, so I think it would only increase noise if we were to add another dropout there (I also did not see an additional dropout on the output for the haiku implementation linked in the issue to add dropout).

**Review by iislucas — COMMENTED** [review]
Looks great, a few small things.

File: animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts
Comment (aliciafmachado) [reply animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts:97]: Done.

File: animated-transformer/src/lib/transformer/transformer_gtensor.ts
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/transformer_gtensor.ts:315]: I suggest we use the T5 implementation in Jax (T5X): https://github.com/google-research/t5x/blob/705247b743d26a33d0c058b41c72ad030e51891b/t5x/examples/t5/network.py#L222

Comment (iislucas) [reply animated-transformer/src/lib/transformer/transformer_gtensor.ts:315]: sg!

**Comment by aliciafmachado** [comment]
Will rebase once #36 is submitted and then pass a generator so that the dropout is reproducible, and then you can take a second look @iislucas.

File: animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts
Comment (iislucas) [inline animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts:81]: remove logs now.

File: animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts
Comment (iislucas) [inline animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts:96]: You fixed this, no?

File: animated-transformer/src/lib/transformer/dropout.ts
Comment (iislucas) [inline animated-transformer/src/lib/transformer/dropout.ts:27]: the type of dimNames should be G[] (or D) I think, since is the set of dimension names (this will catch you accidentally trying to dropout on a non-existent dimension). I think in this case, adding the D extends G doesn't actually help you, so directly using G[] for dimNames will be fine.

**Review by iislucas — COMMENTED** [review]
Looks great, just a couple of minor things.

**Comment by aliciafmachado** [comment]
I added the generator everywhere to be able to seed dropout. I also fixed your comments and small bug. Please take a look.
Additionally, I realized two things:
1. I don't know if it makes sense to have a test with 0.99 dropout because one single SGD step is highly sensitive and even for the case without dropout, if you change the seed, the loss does not necessarily go down. So I think that we can either remove the test with dropout or just test running a sgd step on it to make sure nothing is broken.
2. If the dropout is very high (0.999) sometimes you get a NaN loss, I think it probably has to do with everything being dropped out - but I need to investigate a bit more. I don't think we will use such high dropout in any case. So I think I can create an issue to investigate this corner case, and we can unblock the PR if you agree.

File: animated-transformer/src/lib/transformer/dropout.ts
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/dropout.ts:27]: Done

File: animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts
Comment (aliciafmachado) [reply animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts]: Now it's fixed :)

File: animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts
Comment (aliciafmachado) [reply animated-transformer/src/lib/trainer/basic_transformer_trainer.spec.ts]: done

**Comment by aliciafmachado** [comment]
There's one additional thing: ideally we should just pass one dropout rate and use it everywhere, but the way it's setup we have to pass in each layer and even outside. I can try to think of a way to make it clean and just pass it once, but I can also open the issue and fix it in a future PR.

**Review by iislucas — APPROVED** [review]
Thanks! LGTM. We can discuss improvements more later, but this is great.

**Pull request merged**
